### ISSUE TITLE
`getattr` does not perform type narrowing

### ISSUE THREAD

**Issue by S-D-R** [microsoft/pyright#6381]
**Describe the bug**

When calling `getattr` on an object which does not define a custom `__getattr__`, using a literal string as the key, I would expect pyright  to treat this in the same way as a regular attribute access for the purposes of type narrowing.

**Code or Screenshots**
```python
from dataclasses import dataclass
from typing import Optional


@dataclass
class Foo:
    x: Optional[str]


# typechecks
def f1(foo: Foo) -> str:
    if foo.x is None:
        return "None"
    return foo.x


# does not typecheck
def f2(foo: Foo) -> str:
    if getattr(foo, "x") is None:
        return "None"
    return foo.x  # Type "None" cannot be assigned to type "str"
```

This might look a bit contrived, so here is something a bit closer to my actual use case:

```python
import sys


def f(o: object) -> str:
    m = sys.modules[o.__module__]

    # m.__file__ is typed as "str | None", but when running in a REPL or notebook, the __file__ attribute
    # is actually not defined on the __main__ module, hence the getattr with a default value
    if getattr(m, "__file__", None) is None:
        return "__main__"

    return m.__file__  # Type "None" cannot be assigned to type "str"
```

**VS Code extension or command-line**
Both.

**Comment by erictraut** [comment]
Yes, this is correct. The `getattr` call is not a supported type guard. For a full list of supported type guards, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

We occasionally add support for new type guard patterns, but each one requires specific code that needs to be added and maintained. They can also have a negative impact on overall type analysis performance, so the bar is high for adding new forms. Your sample above actually uses two different forms (`getattr(x, <Literal>) is None` and `getattr(x, <Literal>, None) is None`), and both of those would need to be added separately if we were to support them. It's unlikely we'd do so.

Since `Foo.x` is already declared to contain either `str` or `None`, you should not need to use `getattr` here. You can simply test `if foo.x is None`. This is a supported type guard form.

If you need to use more complex forms, you can leverage `TypeGuard` (see [PEP 647](https://peps.python.org/pep-0647/) for details).
